### PR TITLE
Add a fine-tune command for continuing training on a new dataset

### LINKED ISSUE #879: tuning a pre-trained model
Based on the current instructions I am able to train the system (from scratch). 
Is it possible to load an existing model and continue training it for a few iterations on another data? 

(I briefly asked @OyvindTafjord about this; I know he has hacked his way in to get this, on a relatively older branch. Wondering if there is a better to accomplish this in the newer versions)

### PR DESCRIPTION
**Pull request by matt-gardner**
Fixes #879.

I'm sure that we'll want to add more to this eventually, but this _should_ be the minimum functionality needed for continuing training a model on a new dataset.

I based this off of the `train` command, and I cleaned up some of what was there as I went, so there are also some minor changes to existing files.

@danyaljj, FYI.

I'm "requesting a review" now.  I'm working on the weekend because my family is gone for a few hours this morning, and I wanted to get this done.  I sometimes time-shift my hours a bit so that I can help with family stuff during the week.  _Please_ don't anyone feel pressure to work on the weekends just because I am doing this now.

### CODE REVIEW

File: allennlp/commands/__init__.py [hunk 1]
```diff
@@ -43,6 +47,7 @@ def main(prog: str = None,
             "serve": Serve(model_overrides),
             "make-vocab": MakeVocab(),
             "elmo": Elmo(),
+            "fine-tune": FineTune(),
```
Comment (danyaljj) [inline allennlp/commands/__init__.py]: I'm trying it now. 
An early comment: Would be nice to have a way to set the number of epochs.

Comment (danyaljj) [reply allennlp/commands/__init__.py]: Actually, nvm. The goal of `overrides` is to let people set such extra params. 
That said, would be nice to give a few examples of overrides. Not trivial how to do so (what format? What keys are available? etc)

Comment (danyaljj) [reply allennlp/commands/__init__.py]: An example struggle in my case: 
I tried all of these: 
```
-o "num_epochs": 5 
-o { "num_epochs": 5 } 
-o "'num_epochs': 5 "
-o "num_epochs: 5 "
```

Until I realized this is the right syntax: 
```
-o "trainer: { num_epochs: 5 }"
```

File: allennlp/commands/fine_tune.py [hunk 2]
```diff
@@ -0,0 +1,202 @@
...
+    ----------
+    model_archive_path : ``str``
+        Path to a saved model archive that is the result of running the ``train`` command.
+    data_path : ``str``
+        The training data to use for continuing training the saved model.  We just pass this along
+        to :func:`fine_tune_model`.
+    serialization_dir : ``str``
+        The directory in which to save results and logs. We just pass this along to
+        :func:`fine_tune_model`.
+    validation_data_path : ``str``, optional
```
Comment (danyaljj) [inline allennlp/commands/fine_tune.py]: While validation is optional, without it, I get this error: 

```
Traceback (most recent call last):
  File "/usr/local/Cellar/python3/3.6.4_2/Frameworks/Python.framework/Versions/3.6/lib/python3.6/runpy.py", line 193, in _run_module_as_main
    "__main__", mod_spec)
  File "/usr/local/Cellar/python3/3.6.4_2/Frameworks/Python.framework/Versions/3.6/lib/python3.6/runpy.py", line 85, in _run_code
    exec(code, run_globals)
  File "/Users/daniel/ideaProjects/allennlp/allennlp/run.py", line 18, in <module>
    main(prog="python -m allennlp.run")
  File "/Users/daniel/ideaProjects/allennlp/allennlp/commands/__init__.py", line 65, in main
    args.func(args)
  File "/Users/daniel/ideaProjects/allennlp/allennlp/commands/fine_tune.py", line 87, in fine_tune_model_from_args
    file_friendly_logging=args.file_friendly_logging)
  File "/Users/daniel/ideaProjects/allennlp/allennlp/commands/fine_tune.py", line 125, in fine_tune_model_from_file_paths
    file_friendly_logging=file_friendly_logging)
  File "/Users/daniel/ideaProjects/allennlp/allennlp/commands/fine_tune.py", line 192, in fine_tune_model
    metrics = trainer.train()
  File "/Users/daniel/ideaProjects/allennlp/allennlp/training/trainer.py", line 673, in train
    self._update_learning_rate(epoch, val_metric=this_epoch_val_metric)
  File "/Users/daniel/ideaProjects/allennlp/allennlp/training/trainer.py", line 591, in _update_learning_rate
    raise ConfigurationError("The reduce_on_plateau learning rate scheduler requires "
allennlp.common.checks.ConfigurationError: 'The reduce_on_plateau learning rate scheduler requires a validation metric to compute the schedule and therefore must be used with a validation dataset.'
```

Comment (danyaljj) [reply allennlp/commands/fine_tune.py]: Related to this: I am having some issues setting the validation set from commandline. Gave up trying and hard-coded it in the code.

**Comment by danyaljj** [comment]
(Not sure if it is directly related to the changes here) At some point I got this error: 
```
  File "/Users/daniel/ideaProjects/allennlp/allennlp/common/params.py", line 246, in assert_empty
    raise ConfigurationError("Extra parameters passed to {}: {}".format(class_name, self.params))
allennlp.common.checks.ConfigurationError: "Extra parameters passed to Trainer: ConfigTree([('no_tqdm', True)])" 
```

The complete output: 
```
huntsman-ve507-0729:allennlp daniel$  python3.6 -m allennlp.run fine-tune --model-archive https://s3-us-west-2.amazonaws.com/allennlp/models/bidaf-model-2017.09.15-charpad.tar.gz  -d /Users/daniel/Desktop/train_all_RemediaOnly_squadFormat.json -s  /Users/daniel/Dropbox/bidaf_train/bidaf-fine-tuned-iter=2 -o "trainer: { num_epochs: 2, cuda_device: -1, learning_rate_scheduler: { factor: 0.005 } }" 

/usr/local/lib/python3.6/site-packages/h5py/__init__.py:36: FutureWarning: Conversion of the second argument of issubdtype from `float` to `np.floating` is deprecated. In future, it will be treated as `np.float64 == np.dtype(float).type`.
  from ._conv import register_converters as _register_converters
2018-02-18 22:44:57,970 - INFO - allennlp.models.archival - extracting archive file /Users/daniel/.allennlp/datasets/aHR0cHM6Ly9zMy11cy13ZXN0LTIuYW1hem9uYXdzLmNvbS9hbGxlbm5scC9tb2RlbHMvYmlkYWYtbW9kZWwtMjAxNy4wOS4xNS1jaGFycGFkLnRhci5neg==.bc6cc08ed491c6ac831bf363f7312fc1-6 to temp dir /var/folders/g9/72x1k9_s6sjgm_h5z_g4dwc80000gn/T/tmpb0g9jomr
2018-02-18 22:44:58,511 - INFO - allennlp.data.vocabulary - Loading token dictionary from /var/folders/g9/72x1k9_s6sjgm_h5z_g4dwc80000gn/T/tmpb0g9jomr/vocabulary.
2018-02-18 22:44:58,645 - INFO - allennlp.common.params - model.type = bidaf
2018-02-18 22:44:58,645 - INFO - allennlp.common.params - model.text_field_embedder.type = basic
2018-02-18 22:44:58,646 - INFO - allennlp.common.params - model.text_field_embedder.tokens.type = embedding
2018-02-18 22:44:58,646 - INFO - allennlp.common.params - model.text_field_embedder.tokens.num_embeddings = None
2018-02-18 22:44:58,646 - INFO - allennlp.common.params - model.text_field_embedder.tokens.vocab_namespace = tokens
2018-02-18 22:44:58,646 - INFO - allennlp.common.params - model.text_field_embedder.tokens.embedding_dim = 100
2018-02-18 22:44:58,646 - INFO - allennlp.common.params - model.text_field_embedder.tokens.pretrained_file = None
2018-02-18 22:44:58,646 - INFO - allennlp.common.params - model.text_field_embedder.tokens.projection_dim = None
2018-02-18 22:44:58,646 - INFO - allennlp.common.params - model.text_field_embedder.tokens.trainable = False
2018-02-18 22:44:58,646 - INFO - allennlp.common.params - model.text_field_embedder.tokens.padding_index = None
2018-02-18 22:44:58,646 - INFO - allennlp.common.params - model.text_field_embedder.tokens.max_norm = None
2018-02-18 22:44:58,646 - INFO - allennlp.common.params - model.text_field_embedder.tokens.norm_type = 2.0
2018-02-18 22:44:58,646 - INFO - allennlp.common.params - model.text_field_embedder.tokens.scale_grad_by_freq = False
2018-02-18 22:44:58,646 - INFO - allennlp.common.params - model.text_field_embedder.tokens.sparse = False
2018-02-18 22:44:59,142 - INFO - allennlp.common.params - model.text_field_embedder.token_characters.type = character_encoding
2018-02-18 22:44:59,142 - INFO - allennlp.common.params - model.text_field_embedder.token_characters.embedding.num_embeddings = 262
2018-02-18 22:44:59,142 - INFO - allennlp.common.params - model.text_field_embedder.token_characters.embedding.vocab_namespace = token_characters
2018-02-18 22:44:59,142 - INFO - allennlp.common.params - model.text_field_embedder.token_characters.embedding.embedding_dim = 16
2018-02-18 22:44:59,142 - INFO - allennlp.common.params - model.text_field_embedder.token_characters.embedding.pretrained_file = None
2018-02-18 22:44:59,142 - INFO - allennlp.common.params - model.text_field_embedder.token_characters.embedding.projection_dim = None
2018-02-18 22:44:59,142 - INFO - allennlp.common.params - model.text_field_embedder.token_characters.embedding.trainable = True
2018-02-18 22:44:59,142 - INFO - allennlp.common.params - model.text_field_embedder.token_characters.embedding.padding_index = None
2018-02-18 22:44:59,142 - INFO - allennlp.common.params - model.text_field_embedder.token_characters.embedding.max_norm = None
2018-02-18 22:44:59,142 - INFO - allennlp.common.params - model.text_field_embedder.token_characters.embedding.norm_type = 2.0
2018-02-18 22:44:59,143 - INFO - allennlp.common.params - model.text_field_embedder.token_characters.embedding.scale_grad_by_freq = False
2018-02-18 22:44:59,143 - INFO - allennlp.common.params - model.text_field_embedder.token_characters.embedding.sparse = False
2018-02-18 22:44:59,143 - INFO - allennlp.common.params - model.text_field_embedder.token_characters.encoder.type = cnn
2018-02-18 22:44:59,143 - INFO - allennlp.common.params - model.text_field_embedder.token_characters.encoder.embedding_dim = 16
2018-02-18 22:44:59,143 - INFO - allennlp.common.params - model.text_field_embedder.token_characters.encoder.output_dim = None
2018-02-18 22:44:59,143 - INFO - allennlp.common.params - model.text_field_embedder.token_characters.encoder.num_filters = 100
2018-02-18 22:44:59,143 - INFO - allennlp.common.params - model.text_field_embedder.token_characters.encoder.conv_layer_activation = relu
2018-02-18 22:44:59,144 - INFO - allennlp.common.params - model.text_field_embedder.token_characters.encoder.ngram_filter_sizes = [5]
2018-02-18 22:44:59,145 - INFO - allennlp.common.params - model.text_field_embedder.token_characters.dropout = 0.2
2018-02-18 22:44:59,146 - INFO - allennlp.common.params - model.num_highway_layers = 2
2018-02-18 22:44:59,146 - INFO - allennlp.common.params - model.phrase_layer.type = lstm
2018-02-18 22:44:59,146 - INFO - allennlp.common.params - model.phrase_layer.batch_first = True
2018-02-18 22:44:59,146 - INFO - allennlp.common.params - Converting Params object to dict; logging of default values will not occur when dictionary parameters are used subsequently.
2018-02-18 22:44:59,146 - INFO - allennlp.common.params - CURRENTLY DEFINED PARAMETERS: 
2018-02-18 22:44:59,146 - INFO - allennlp.common.params - model.phrase_layer.bidirectional = True
2018-02-18 22:44:59,146 - INFO - allennlp.common.params - model.phrase_layer.input_size = 200
2018-02-18 22:44:59,146 - INFO - allennlp.common.params - model.phrase_layer.hidden_size = 100
2018-02-18 22:44:59,146 - INFO - allennlp.common.params - model.phrase_layer.num_layers = 1
2018-02-18 22:44:59,146 - INFO - allennlp.common.params - model.phrase_layer.dropout = 0.2
2018-02-18 22:44:59,146 - INFO - allennlp.common.params - model.phrase_layer.batch_first = True
2018-02-18 22:44:59,148 - INFO - allennlp.common.params - model.similarity_function.type = linear
2018-02-18 22:44:59,148 - INFO - allennlp.common.params - model.similarity_function.tensor_1_dim = 200
2018-02-18 22:44:59,148 - INFO - allennlp.common.params - model.similarity_function.tensor_2_dim = 200
2018-02-18 22:44:59,149 - INFO - allennlp.common.params - model.similarity_function.combination = x,y,x*y
2018-02-18 22:44:59,149 - INFO - allennlp.common.params - model.similarity_function.activation = linear
2018-02-18 22:44:59,149 - INFO - allennlp.common.params - model.modeling_layer.type = lstm
2018-02-18 22:44:59,149 - INFO - allennlp.common.params - model.modeling_layer.batch_first = True
2018-02-18 22:44:59,149 - INFO - allennlp.common.params - Converting Params object to dict; logging of default values will not occur when dictionary parameters are used subsequently.
2018-02-18 22:44:59,149 - INFO - allennlp.common.params - CURRENTLY DEFINED PARAMETERS: 
2018-02-18 22:44:59,149 - INFO - allennlp.common.params - model.modeling_layer.bidirectional = True
2018-02-18 22:44:59,149 - INFO - allennlp.common.params - model.modeling_layer.input_size = 800
2018-02-18 22:44:59,149 - INFO - allennlp.common.params - model.modeling_layer.hidden_size = 100
2018-02-18 22:44:59,149 - INFO - allennlp.common.params - model.modeling_layer.num_layers = 2
2018-02-18 22:44:59,149 - INFO - allennlp.common.params - model.modeling_layer.dropout = 0.2
2018-02-18 22:44:59,149 - INFO - allennlp.common.params - model.modeling_layer.batch_first = True
2018-02-18 22:44:59,158 - INFO - allennlp.common.params - model.span_end_encoder.type = lstm
2018-02-18 22:44:59,158 - INFO - allennlp.common.params - model.span_end_encoder.batch_first = True
2018-02-18 22:44:59,158 - INFO - allennlp.common.params - Converting Params object to dict; logging of default values will not occur when dictionary parameters are used subsequently.
2018-02-18 22:44:59,158 - INFO - allennlp.common.params - CURRENTLY DEFINED PARAMETERS: 
2018-02-18 22:44:59,158 - INFO - allennlp.common.params - model.span_end_encoder.bidirectional = True
2018-02-18 22:44:59,158 - INFO - allennlp.common.params - model.span_end_encoder.input_size = 1400
2018-02-18 22:44:59,158 - INFO - allennlp.common.params - model.span_end_encoder.hidden_size = 100
2018-02-18 22:44:59,158 - INFO - allennlp.common.params - model.span_end_encoder.num_layers = 1
2018-02-18 22:44:59,158 - INFO - allennlp.common.params - model.span_end_encoder.dropout = 0.2
2018-02-18 22:44:59,158 - INFO - allennlp.common.params - model.span_end_encoder.batch_first = True
2018-02-18 22:44:59,167 - INFO - allennlp.common.params - model.dropout = 0.2
2018-02-18 22:44:59,167 - INFO - allennlp.common.params - model.initializer = []
2018-02-18 22:44:59,167 - INFO - allennlp.common.params - model.regularizer = []
2018-02-18 22:44:59,167 - INFO - allennlp.common.params - model.mask_lstms = True
2018-02-18 22:44:59,170 - INFO - allennlp.nn.initializers - Initializing parameters
2018-02-18 22:44:59,170 - INFO - allennlp.nn.initializers - Done initializing parameters; the following parameters are using their default initialization from their code
2018-02-18 22:44:59,170 - INFO - allennlp.nn.initializers -    _highway_layer._module._layers.0.bias
2018-02-18 22:44:59,170 - INFO - allennlp.nn.initializers -    _highway_layer._module._layers.0.weight
2018-02-18 22:44:59,170 - INFO - allennlp.nn.initializers -    _highway_layer._module._layers.1.bias
2018-02-18 22:44:59,170 - INFO - allennlp.nn.initializers -    _highway_layer._module._layers.1.weight
2018-02-18 22:44:59,170 - INFO - allennlp.nn.initializers -    _matrix_attention._similarity_function._bias
2018-02-18 22:44:59,170 - INFO - allennlp.nn.initializers -    _matrix_attention._similarity_function._weight_vector
2018-02-18 22:44:59,170 - INFO - allennlp.nn.initializers -    _modeling_layer._module.bias_hh_l0
2018-02-18 22:44:59,170 - INFO - allennlp.nn.initializers -    _modeling_layer._module.bias_hh_l0_reverse
2018-02-18 22:44:59,171 - INFO - allennlp.nn.initializers -    _modeling_layer._module.bias_hh_l1
2018-02-18 22:44:59,171 - INFO - allennlp.nn.initializers -    _modeling_layer._module.bias_hh_l1_reverse
2018-02-18 22:44:59,171 - INFO - allennlp.nn.initializers -    _modeling_layer._module.bias_ih_l0
2018-02-18 22:44:59,171 - INFO - allennlp.nn.initializers -    _modeling_layer._module.bias_ih_l0_reverse
2018-02-18 22:44:59,171 - INFO - allennlp.nn.initializers -    _modeling_layer._module.bias_ih_l1
2018-02-18 22:44:59,171 - INFO - allennlp.nn.initializers -    _modeling_layer._module.bias_ih_l1_reverse
2018-02-18 22:44:59,171 - INFO - allennlp.nn.initializers -    _modeling_layer._module.weight_hh_l0
2018-02-18 22:44:59,171 - INFO - allennlp.nn.initializers -    _modeling_layer._module.weight_hh_l0_reverse
2018-02-18 22:44:59,171 - INFO - allennlp.nn.initializers -    _modeling_layer._module.weight_hh_l1
2018-02-18 22:44:59,171 - INFO - allennlp.nn.initializers -    _modeling_layer._module.weight_hh_l1_reverse
2018-02-18 22:44:59,171 - INFO - allennlp.nn.initializers -    _modeling_layer._module.weight_ih_l0
2018-02-18 22:44:59,171 - INFO - allennlp.nn.initializers -    _modeling_layer._module.weight_ih_l0_reverse
2018-02-18 22:44:59,171 - INFO - allennlp.nn.initializers -    _modeling_layer._module.weight_ih_l1
2018-02-18 22:44:59,171 - INFO - allennlp.nn.initializers -    _modeling_layer._module.weight_ih_l1_reverse
2018-02-18 22:44:59,171 - INFO - allennlp.nn.initializers -    _phrase_layer._module.bias_hh_l0
2018-02-18 22:44:59,171 - INFO - allennlp.nn.initializers -    _phrase_layer._module.bias_hh_l0_reverse
2018-02-18 22:44:59,171 - INFO - allennlp.nn.initializers -    _phrase_layer._module.bias_ih_l0
2018-02-18 22:44:59,171 - INFO - allennlp.nn.initializers -    _phrase_layer._module.bias_ih_l0_reverse
2018-02-18 22:44:59,171 - INFO - allennlp.nn.initializers -    _phrase_layer._module.weight_hh_l0
2018-02-18 22:44:59,171 - INFO - allennlp.nn.initializers -    _phrase_layer._module.weight_hh_l0_reverse
2018-02-18 22:44:59,171 - INFO - allennlp.nn.initializers -    _phrase_layer._module.weight_ih_l0
2018-02-18 22:44:59,172 - INFO - allennlp.nn.initializers -    _phrase_layer._module.weight_ih_l0_reverse
2018-02-18 22:44:59,172 - INFO - allennlp.nn.initializers -    _span_end_encoder._module.bias_hh_l0
2018-02-18 22:44:59,172 - INFO - allennlp.nn.initializers -    _span_end_encoder._module.bias_hh_l0_reverse
2018-02-18 22:44:59,172 - INFO - allennlp.nn.initializers -    _span_end_encoder._module.bias_ih_l0
2018-02-18 22:44:59,172 - INFO - allennlp.nn.initializers -    _span_end_encoder._module.bias_ih_l0_reverse
2018-02-18 22:44:59,172 - INFO - allennlp.nn.initializers -    _span_end_encoder._module.weight_hh_l0
2018-02-18 22:44:59,172 - INFO - allennlp.nn.initializers -    _span_end_encoder._module.weight_hh_l0_reverse
2018-02-18 22:44:59,172 - INFO - allennlp.nn.initializers -    _span_end_encoder._module.weight_ih_l0
2018-02-18 22:44:59,172 - INFO - allennlp.nn.initializers -    _span_end_encoder._module.weight_ih_l0_reverse
2018-02-18 22:44:59,172 - INFO - allennlp.nn.initializers -    _span_end_predictor._module.bias
2018-02-18 22:44:59,172 - INFO - allennlp.nn.initializers -    _span_end_predictor._module.weight
2018-02-18 22:44:59,172 - INFO - allennlp.nn.initializers -    _span_start_predictor._module.bias
2018-02-18 22:44:59,172 - INFO - allennlp.nn.initializers -    _span_start_predictor._module.weight
2018-02-18 22:44:59,172 - INFO - allennlp.nn.initializers -    _text_field_embedder.token_embedder_token_characters._embedding._module.weight
2018-02-18 22:44:59,172 - INFO - allennlp.nn.initializers -    _text_field_embedder.token_embedder_token_characters._encoder._module.conv_layer_0.bias
2018-02-18 22:44:59,172 - INFO - allennlp.nn.initializers -    _text_field_embedder.token_embedder_token_characters._encoder._module.conv_layer_0.weight
2018-02-18 22:44:59,172 - INFO - allennlp.nn.initializers -    _text_field_embedder.token_embedder_tokens.weight
2018-02-18 22:44:59,243 - INFO - allennlp.common.params - random_seed = 13370
2018-02-18 22:44:59,243 - INFO - allennlp.common.params - numpy_seed = 1337
2018-02-18 22:44:59,243 - INFO - allennlp.common.params - pytorch_seed = 133
2018-02-18 22:44:59,244 - INFO - allennlp.common.checks - Pytorch version: 0.3.0.post4
2018-02-18 22:44:59,601 - INFO - allennlp.common.params - iterator.type = bucket
2018-02-18 22:44:59,601 - INFO - allennlp.common.params - iterator.sorting_keys = [['passage', 'num_tokens'], ['question', 'num_tokens']]
2018-02-18 22:44:59,601 - INFO - allennlp.common.params - iterator.padding_noise = 0.1
2018-02-18 22:44:59,601 - INFO - allennlp.common.params - iterator.biggest_batch_first = False
2018-02-18 22:44:59,602 - INFO - allennlp.common.params - iterator.batch_size = 40
2018-02-18 22:44:59,602 - INFO - allennlp.common.params - iterator.instances_per_epoch = None
2018-02-18 22:44:59,602 - INFO - allennlp.common.params - iterator.max_instances_in_memory = None
2018-02-18 22:44:59,602 - INFO - allennlp.common.params - dataset_reader.type = squad
2018-02-18 22:44:59,602 - INFO - allennlp.common.params - dataset_reader.tokenizer.type = word
2018-02-18 22:44:59,603 - INFO - allennlp.common.params - dataset_reader.tokenizer.word_splitter.type = spacy
2018-02-18 22:44:59,603 - INFO - allennlp.common.params - dataset_reader.tokenizer.word_splitter.language = en_core_web_sm
2018-02-18 22:44:59,603 - INFO - allennlp.common.params - dataset_reader.tokenizer.word_splitter.pos_tags = False
2018-02-18 22:44:59,603 - INFO - allennlp.common.params - dataset_reader.tokenizer.word_splitter.parse = False
2018-02-18 22:44:59,603 - INFO - allennlp.common.params - dataset_reader.tokenizer.word_splitter.ner = False
2018-02-18 22:45:00,188 - INFO - allennlp.common.params - dataset_reader.tokenizer.word_filter.type = pass_through
2018-02-18 22:45:00,188 - INFO - allennlp.common.params - dataset_reader.tokenizer.word_stemmer.type = pass_through
2018-02-18 22:45:00,188 - INFO - allennlp.common.params - dataset_reader.tokenizer.start_tokens = None
2018-02-18 22:45:00,188 - INFO - allennlp.common.params - dataset_reader.tokenizer.end_tokens = None
2018-02-18 22:45:00,189 - INFO - allennlp.common.params - dataset_reader.token_indexers.tokens.type = single_id
2018-02-18 22:45:00,189 - INFO - allennlp.common.params - dataset_reader.token_indexers.tokens.namespace = tokens
2018-02-18 22:45:00,189 - INFO - allennlp.common.params - dataset_reader.token_indexers.tokens.lowercase_tokens = True
2018-02-18 22:45:00,189 - INFO - allennlp.common.params - dataset_reader.token_indexers.token_characters.type = characters
2018-02-18 22:45:00,189 - INFO - allennlp.common.params - dataset_reader.token_indexers.token_characters.namespace = token_characters
2018-02-18 22:45:00,189 - INFO - allennlp.common.params - dataset_reader.token_indexers.token_characters.character_tokenizer.byte_encoding = utf-8
2018-02-18 22:45:00,190 - INFO - allennlp.common.params - dataset_reader.token_indexers.token_characters.character_tokenizer.lowercase_characters = False
2018-02-18 22:45:00,190 - INFO - allennlp.common.params - dataset_reader.token_indexers.token_characters.character_tokenizer.start_tokens = [259]
2018-02-18 22:45:00,190 - INFO - allennlp.common.params - dataset_reader.token_indexers.token_characters.character_tokenizer.end_tokens = [260, 0, 0, 0, 0, 0]
2018-02-18 22:45:00,190 - INFO - allennlp.common.params - dataset_reader.lazy = False
2018-02-18 22:45:00,190 - INFO - allennlp.common.params - validation_dataset_reader = None
2018-02-18 22:45:00,190 - INFO - allennlp.common.params - train_data_path = /Users/daniel/Desktop/train_all_RemediaOnly_squadFormat.json
2018-02-18 22:45:00,191 - INFO - allennlp.commands.train - Reading training data from /Users/daniel/Desktop/train_all_RemediaOnly_squadFormat.json
0it [00:00, ?it/s]2018-02-18 22:45:00,192 - INFO - allennlp.data.dataset_readers.reading_comprehension.squad - Reading file at /Users/daniel/Desktop/train_all_RemediaOnly_squadFormat.json
2018-02-18 22:45:00,194 - INFO - allennlp.data.dataset_readers.reading_comprehension.squad - Reading the dataset
272it [00:00, 645.81it/s]
2018-02-18 22:45:00,613 - INFO - allennlp.common.params - validation_data_path = /Users/daniel/Desktop/train_all_RemediaOnly_squadFormat.json
2018-02-18 22:45:00,614 - INFO - allennlp.commands.train - Reading validation data from /Users/daniel/Desktop/train_all_RemediaOnly_squadFormat.json
0it [00:00, ?it/s]2018-02-18 22:45:00,614 - INFO - allennlp.data.dataset_readers.reading_comprehension.squad - Reading file at /Users/daniel/Desktop/train_all_RemediaOnly_squadFormat.json
2018-02-18 22:45:00,615 - INFO - allennlp.data.dataset_readers.reading_comprehension.squad - Reading the dataset
272it [00:00, 1050.84it/s]
2018-02-18 22:45:00,873 - INFO - allennlp.common.params - test_data_path = None
2018-02-18 22:45:00,874 - INFO - allennlp.common.params - trainer.patience = 10
2018-02-18 22:45:00,874 - INFO - allennlp.common.params - trainer.validation_metric = +em
2018-02-18 22:45:00,874 - INFO - allennlp.common.params - trainer.num_epochs = 2
2018-02-18 22:45:00,875 - INFO - allennlp.common.params - trainer.cuda_device = -1
2018-02-18 22:45:00,875 - INFO - allennlp.common.params - trainer.grad_norm = 5.0
2018-02-18 22:45:00,875 - INFO - allennlp.common.params - trainer.grad_clipping = None
2018-02-18 22:45:00,875 - INFO - allennlp.common.params - trainer.optimizer.type = adam
2018-02-18 22:45:00,875 - INFO - allennlp.common.params - trainer.optimizer.parameter_groups = None
2018-02-18 22:45:00,876 - INFO - allennlp.common.params - Converting Params object to dict; logging of default values will not occur when dictionary parameters are used subsequently.
2018-02-18 22:45:00,876 - INFO - allennlp.common.params - CURRENTLY DEFINED PARAMETERS: 
2018-02-18 22:45:00,876 - INFO - allennlp.common.params - trainer.optimizer.betas = [0.9, 0.9]
2018-02-18 22:45:00,876 - INFO - allennlp.common.params - trainer.learning_rate_scheduler.type = reduce_on_plateau
2018-02-18 22:45:00,876 - INFO - allennlp.common.params - Converting Params object to dict; logging of default values will not occur when dictionary parameters are used subsequently.
2018-02-18 22:45:00,876 - INFO - allennlp.common.params - CURRENTLY DEFINED PARAMETERS: 
2018-02-18 22:45:00,876 - INFO - allennlp.common.params - trainer.learning_rate_scheduler.factor = 0.005
2018-02-18 22:45:00,876 - INFO - allennlp.common.params - trainer.learning_rate_scheduler.mode = max
2018-02-18 22:45:00,877 - INFO - allennlp.common.params - trainer.learning_rate_scheduler.patience = 2
2018-02-18 22:45:00,877 - INFO - allennlp.common.params - trainer.num_serialized_models_to_keep = None
2018-02-18 22:45:00,877 - INFO - allennlp.common.params - trainer.keep_serialized_model_every_num_seconds = None
2018-02-18 22:45:00,877 - INFO - allennlp.common.params - trainer.model_save_interval = None
2018-02-18 22:45:00,877 - INFO - allennlp.common.params - trainer.summary_interval = 100
2018-02-18 22:45:00,877 - INFO - allennlp.common.params - trainer.histogram_interval = None
Traceback (most recent call last):
  File "/usr/local/Cellar/python3/3.6.4_2/Frameworks/Python.framework/Versions/3.6/lib/python3.6/runpy.py", line 193, in _run_module_as_main
    "__main__", mod_spec)
  File "/usr/local/Cellar/python3/3.6.4_2/Frameworks/Python.framework/Versions/3.6/lib/python3.6/runpy.py", line 85, in _run_code
    exec(code, run_globals)
  File "/Users/daniel/ideaProjects/allennlp/allennlp/run.py", line 18, in <module>
    main(prog="python -m allennlp.run")
  File "/Users/daniel/ideaProjects/allennlp/allennlp/commands/__init__.py", line 65, in main
    args.func(args)
  File "/Users/daniel/ideaProjects/allennlp/allennlp/commands/fine_tune.py", line 87, in fine_tune_model_from_args
    file_friendly_logging=args.file_friendly_logging)
  File "/Users/daniel/ideaProjects/allennlp/allennlp/commands/fine_tune.py", line 125, in fine_tune_model_from_file_paths
    file_friendly_logging=file_friendly_logging)
  File "/Users/daniel/ideaProjects/allennlp/allennlp/commands/fine_tune.py", line 191, in fine_tune_model
    trainer_params)
  File "/Users/daniel/ideaProjects/allennlp/allennlp/training/trainer.py", line 887, in from_params
    params.assert_empty(cls.__name__)
  File "/Users/daniel/ideaProjects/allennlp/allennlp/common/params.py", line 246, in assert_empty
    raise ConfigurationError("Extra parameters passed to {}: {}".format(class_name, self.params))
allennlp.common.checks.ConfigurationError: "Extra parameters passed to Trainer: ConfigTree([('no_tqdm', True)])"```

**Comment by danyaljj** [comment]
Just to confirm that things work: I finally managed to tune a model on data! :)

**Comment by maoredman** [comment]
@danyaljj I think the `"Extra parameters passed to Trainer: ConfigTree([('no_tqdm', True)])"` error was caused by the removal of `params.pop_bool("no_tqdm", False)` from the `Trainer` class

https://github.com/allenai/allennlp/commit/2a90ed83ac62a3ab77bb8eb9d4c1543b73df729c#diff-043dcd121296c3cef3f3ff8c74127ff1L543

File: allennlp/commands/fine_tune.py (hunk 2 above)
Comment (joelgrus) [reply allennlp/commands/fine_tune.py]: FYI: validation is optional in general, but your experiment seems to be using the `reduce_on_plateau` learning rate scheduler, which requires a validation set to work

Comment (danyaljj) [reply allennlp/commands/fine_tune.py]: Ya that's what I figured after waiting ~20mins for to see this error. Ideally, the system should be smart to get this (say, with a clear error msg, right at the beginning).

Comment (matt-gardner) [reply allennlp/commands/fine_tune.py]: Unfortunately, that's a harder one for us to catch - we don't know what learning rate scheduler you're using, or whether it depends on the validation data.

For the problem setting the validation data from the command line, do you have a specific error?

Comment (danyaljj) [reply allennlp/commands/fine_tune.py]: > Unfortunately, that's a harder one for us to catch - we don't know what learning rate scheduler you're using, or whether it depends on the validation data. 

Hmm ... I was assuming that this should be easy to do. If not, nvm.

File: allennlp/commands/__init__.py [hunk 3]
```diff
@@ -3,35 +3,39 @@
...
-    The default models for ``serve`` and the default predictors for ``predict`` are
-    defined above. If you'd like to add more or use different ones, the
-    ``model_overrides`` and ``predictor_overrides`` arguments will take precedence over the defaults.
+    The :mod:`~allennlp.run` command only knows about the registered classes in the ``allennlp``
+    codebase. In particular, once you start creating your own ``Model`` s and so forth, it won't
+    work for them, unless you use the ``--include-package`` flag available for most commands.
+
+    The default models for ``serve`` and the default predictors for ``predict`` are defined above.
+    If you'd like to add more or use different ones, the ``model_overrides`` and
+    ``predictor_overrides`` arguments will take precedence over the defaults.
```
Comment (joelgrus) [inline allennlp/commands/__init__.py]: this is no longer true, they're defined elsewhere.

also, it occurs to me that now that `predict` has a `--predictor` option there's not really any need for predictor overrides, we should get rid of it

**Comment by matt-gardner** [comment]
@joelgrus this should be ready for another look.

File: allennlp/commands/__init__.py (hunk 3 above)
Comment (matt-gardner) [reply allennlp/commands/__init__.py]: Yeah, you're right.  I added a TODO to take care of the argument fixes later - I think we can probably also remove the `model_overrides`, and maybe just replace the `serve` command with the simple server.

**Comment by saadi143** [comment]
Is there any proper example? which shows the fine-tuned of pre-trained Elmo model. I try a lot but I didn't find any proper example. I don't know how can I fine-tuned the model on my dataset. I am using TensorFlow hub for Elmo embedding with 1024 dimension.
One more thing I want to get the output of a dimension around 300.

**Pull request merged**